src/firebase/AuthService.ts

```diff
@@ -1,31 +1,39 @@
 import firebase from "firebase/app";
 import "firebase/auth";
 
-class AuthService {
-    provider: firebase.auth.OAuthProvider;
-    accessToken: string | undefined;
-    idToken: string | undefined;
+interface ProfileData {
+    isAdmin: boolean,
+    isSnapper: boolean
+}
 
-    constructor() {
-        this.provider = new firebase.auth.OAuthProvider("microsoft.com");
-        this.provider.setCustomParameters({
-            prompt: process.env.REACT_APP_PROMPT,
-            // Only users from a particular Azure AD tenant to sign into the application
-            tenant: process.env.REACT_APP_TENANT,
-        });
+export async function getOrCreateUserProfile(): Promise<any> {
+    const user = firebase.auth().currentUser;
+    if (!user) {
+        throw new Error('Should be logged in by here!')
     }
-
-    async signIn(successCb: () => void, errorCb: (err: string) => void) {
-        try {
-            await firebase.auth().signInWithPopup(this.provider);
-            // If we get to here the sign up has been successful.
-            successCb();
-        } catch (e) {
-            console.error("sign in error");
-            console.error(e);
-            errorCb("There was an error signing in!");
+    const email = user.email;
+    const db = firebase.firestore();
+    const docRef = db.collection("users").doc(email);
+    const doc = await docRef.get();
+    let profileData: ProfileData;
+    if (doc.exists && doc.data() != null) {
+        profileData = doc.data() as ProfileData;
+    } else {
+        profileData = {
+            isAdmin: false,
+            isSnapper: true
         }
+        db.collection("users").doc(email).set(profileData)
     }
+    return profileData;
```

Comment:
How could you simplify this by abstracting appropriate methods?
Things like `getProfile` or `profileExists`?